### ISSUE TITLE
Add option to set preferred language in editor

### ISSUE THREAD

**Issue by andyprv** [node-red/node-red#2088]
upgrade to docker image 0.20.0 , now almost everthing shown in german ... 

We don't want that :-)    

How to switch back to global language en-US ?

**Comment by knolleary** [comment]
It uses your browsers preferred language setting. We don't provide a way to override that currently.

Please try to use the issue template we provide when raising issues. Thanks.

**Comment by andyprv** [comment]
Thanx , 

I changed the Browser language to english US , but NodeRed still in german mode ...

- Browser Restart
- NodeRed Logout / Login

no change ...

Seems like NodeRed using the  Browser - Install settings .. but they dont change 

```BrowserInfo
....
Language:German
System Language:Not detectable with this browser
User Language:de-de
....
```` 

please provide a language override function in the future ... 

meanwhile we switch back to  "nodered/node-red-docker:0.19.6-v8"

**Comment by knolleary** [comment]
What browser are you using?

Using Chrome, I set German as my preferred language, reloaded the node-red editor and everything is in German. I then changed the browser language preference to Japanese, reloaded the editor and everything is in Japanese. Finally I set it to English and reloaded and it's all back to English.

**Comment by dceejay** [comment]
Which browser are you using ?

**Comment by andyprv** [comment]
Your Browser User Agent String is:

Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) 
Chrome/72.0.3626.121 
Safari/537.36

**Comment by knolleary** [comment]
@andyprv are you trying to say you've tried with Firefox, Chrome and Safari?

How are you setting the language?

**Comment by andyprv** [comment]
No only with Chrome , 

that String is the output from http://mybrowserinfo.com/detail.asp?bhcp=1

``` 
Language:English
System Language:Not detectable with this browser
User Language:de-de
```

OK  now It changed the settings .... because you have to arrage the Language settings in your Browser.

Just  switchen the whole browser to english , will not change the Browser info send to the server.. 


![2019-03-13_100557](https://user-images.githubusercontent.com/36567942/54266394-96816d00-4577-11e9-810f-3d46dfc1207d.png)

**Comment by knolleary** [comment]
Right - we assume you want to use your browser's preferred language when displaying the editor.

We haven't needed a language picker before now because we've only had English and Japanese translations.

I think there are two approaches we could take on this and would appreciate your feedback on which would fit best with your use.

1. have an option in settings.js to force the editor into a particular language. This would apply to every user accessing that instance of node-red. This would be fine for individual users and for hosted environments where the audience is known.

2. have an option in the settings menu within the editor to pick preferred language. That option will be stored with the user preferences so remembered for when they come back.

Option 1 is the easiest as it doesn't drive any UI changes - and it doesn't stop us doing option 2 at a later point.

@HiroyasuNishiyama would appreciate your input - do you see any need from your users to be able to switch the language from within the editor?

**Comment by HiroyasuNishiyama** [comment]
@knolleary our users typically read/write Japanese/English, so currently have no request for changing locales dynamically within the editor.

**Comment by andyprv** [comment]
For me / us option #1  is fine.
But other users in countries , with less english Language use , may prefer option #2

m2c  :-)

**Comment by knolleary** [comment]
Thanks all.

We can do option 1 in a 0.20.x maintenance release quite easily. Option 2 may have to wait until 0.21.

**Comment by tameike000** [comment]
Is there an ETA for that fix? The current forced German translation is pretty pointless if 99,99% of all documentation is in English. 

Anyway, sorry for the maybe harsh comment, but that "feature" just costed me 3 hours to update my different node-red setups and then rolling everything back again.

Thanks for Node-red anyway. I am a big fan!
M

**Comment by hardillb** [comment]
Feel free to contribute documentation translations if the fact it's in English is a problem.

**Comment by jwende** [comment]
Why do you set your browsers preference to ‚de‘ or ‚De-de‘ if you don‘t want German language?
J.

**Comment by tameike000** [comment]
Well, I am a German native and I speak a few languages. But just because I speak them it does not mean I set my system language accordingly.
Anyway, if you do any kind of programming the common language is English. That's fine. 
If some users need everything to be translated, that's fine as well. But translate literally a programming language/IDE and forcing users to stick with that is a bit disturbing. Or would you translate i.e. Python commands into German, French or Esperanto?
 
@hardillb, happy to contribute, but translations for a development environment I don't consider necessary. Development docs will be always most up-to-date in English, so why switching back and forth? (German IDE, checking German docs, realizing they are one iteration behind, checking the english one, ... and so one)
@jwende, I think there is a misunderstanding. I just don't want a development environment to define what language I should use.

**Comment by dceejay** [comment]
As noted this issue has been added to the development workstream. There is no formal ETA as real life and the day job sometimes get in the way. In the meantime the temporary workaround is to set just your browser to prefer english (not the whole system).
(Nuclear option is to find and delete the three locales/de directories `@node-red/editor-client/locales/de, @node-red/nodes/locales/de and @node-red/runtime/locales/de`)

**Comment by tameike000** [comment]
@dceejay highly appreciate your hint with a nuclear fix. 

...and please make sure that real life always has priority. It is anyway more important.

Best.

**Comment by wlauppe** [comment]
Thanks for pointing out the "nuclear" workaround thats a perfect workaround for me, for the time beeing.

**Comment by tameike000** [comment]
Quick update:
Nuclear fix also works within docker container:
cd /usr/src/node-red/node_modules/
cd \@node-red/
rm -r editor-client/locales/de
rm -r nodes/locales/de
rm -r runtime/locales/de

**Comment by jwende** [comment]
.... and docker commit to make ist persistent .....
J.

**Comment by jwende** [comment]
We now have a I18n-German channel on NodeRED Slack - please Contact me if you want to join (channel was create as a private Channel)
J.

**Comment by knolleary** [comment]
I have just pushed an change to the `dev` branch for the 1.0 release.

This adds an option in the user settings dialog to pick the desired editor language. After making a selection and closing the dialog, you must reload the page for it to take effect.

The value of the option is stored in localStorage on the browser - so you if access it from two different browsers, you'll have to set it separately on each. This is because we load the message catalogs before we load the user settings from the runtime. If we stored the value in the runtime, by the time the editor knows the user's choice, it would be too late. Any fix to that would require quite a substantial change to how the editor loads, so unless there is major demand for it, I'm not going to consider it.

![Blank](https://user-images.githubusercontent.com/51083/56743746-19006d80-676f-11e9-9b47-82334cc59490.png)

We still need to consider adding an option to settings.js to set the default language - but the same issues applies regarding the order the editor loads things. So that may not be the quick fix I thought it would be.
